baseplate/experiments/__init__.py

```diff
@@ -0,0 +1,171 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import json
+import logging
+
+from .providers import parse_experiment
+from .. import config
+from .._compat import iteritems
+from ..context import ContextFactory
+from ..events import Event, EventTooLargeError, EventQueueFullError
+from ..file_watcher import FileWatcher, WatchedFileNotAvailableError
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExperimentsContextFactory(ContextFactory):
+    """ Experiment client context factory
+
+    This factory will attach a new :py:class:`baseplate.experiments.Experiments`
+    to an attribute on the :term:`context object`.
+    """
+    def __init__(self, path, event_queue):
+        """ ExperimentContextFactory constructor
+
+        :param str path: Path to the experiment config file.
+        :param baseplate.events.EventQueue event_queue: Event queue used for
+            logging bucketing events to the event pipeline.  You can set this
+            to None to disabled bucketing events entirely.
+        """
+        self._filewatcher = FileWatcher(path, json.load)
+        self._event_queue = event_queue
+
+    def make_object_for_context(self, name, server_span):
+        return Experiments(self._filewatcher, self._event_queue)
+
+
+class Experiments(object):
+    """ Access to experiments with automatic refresh when changed. Handles
+    logging bucketing events to the event pipeline.
+    """
+
+    def __init__(self, config_watcher, event_queue=None):
+        self._config_watcher = config_watcher
+        self._event_queue = event_queue
+        self._already_bucketed = set()
+
+    def _get_config(self, name):
+        try:
+            config_data = self._config_watcher.get_data()
+            return config_data[name]
+        except WatchedFileNotAvailableError:
+            logger.warning("Experiment config file not found")
+            return
+        except KeyError:
+            logger.warning(
+                "Experiment <%r> not found in experiment config",
+                name,
+            )
+            return
+        except TypeError:
+            logger.warning("Could not load experiment config: %r", name)
+            return
+
+    def variant(self, name, bucketing_event_override=None,
+                extra_event_params=None, **kwargs):
+        """ Which variant, if any, is active for the experiment specified by
+        "name".  If a variant is active, a bucketing event will be logged to
+        the event pipeline unless any one of the following conditions are met:
+
+        1. bucketing_event_override is set to False.
+        2. The experiment specified by "name" explicitly disables bucketing
+           events.
+        3. We have already logged a bucketing event for the value specified by
+           experiment.event_cache_key(**kwargs) within the current session.
```

Comment:
"session"? should this be "span" or "request" to make it more clear the duration of the debounce?